File: scope_content/common/content_safety.py
from __future__ import annotations

import re
import unicodedata
from dataclasses import dataclass
from typing import Iterable


BLOCKED_REASON = 'blocked_slur_or_hate_term'

_LEET_TRANSLATION = str.maketrans(
    {
        '0': 'o',
        '1': 'i',
        '3': 'e',
        '4': 'a',
        '5': 's',
        '7': 't',
        '@': 'a',
        '$': 's',
    }
)

_BLOCKED_TERMS = {
    'chink',
    'coon',
    'fag',
    'faggot',
    'gook',
    'kike',
    'kill all',
    'kill yourself',
    'kys',
    'nigga',
    'nigger',
    'rape',
    'retard',
    'spic',
    'tranny',
    'wetback',
    # Harmless sentinel used by tests so assertions do not need to spell out a slur.
    'scope test blocked slur',
}

_TOKEN_BLOCKED_TERMS = {
    term
    for term in _BLOCKED_TERMS
    if ' ' not in term and len(term) > 1
}


@dataclass(frozen=True)
class SafetyResult:
    status: str
    reason: str = ''
    field: str | None = None

    @property
    def clean(self) -> bool:
        return self.status == 'clean'


def normalize_safety_text(value: object) -> str:
    normalized = unicodedata.normalize('NFKD', str(value or ''))
    ascii_text = ''.join(character for character in normalized if not unicodedata.combining(character))
    return ascii_text.casefold().translate(_LEET_TRANSLATION)


def _tokenize(value: object) -> list[str]:
    return [token for token in re.split(r'[^a-z0-9]+', normalize_safety_text(value)) if token]


def _contains_blocked_term(value: object) -> bool:
    tokens = _tokenize(value)
    if not tokens:
        return False

    token_set = set(tokens)
    if token_set.intersection(_TOKEN_BLOCKED_TERMS):
        return True

    compact_phrase = ' '.join(tokens)
    return any(term in compact_phrase for term in _BLOCKED_TERMS if ' ' in term)


def evaluate_text_fields(fields: Iterable[tuple[str, object]]) -> SafetyResult:
    for field, value in fields:
        if isinstance(value, (list, tuple, set)):
            values = value
        else:
            values = [value]

        if any(_contains_blocked_term(item) for item in values):
            return SafetyResult(status='blocked', reason=BLOCKED_REASON, field=field)

    return SafetyResult(status='clean')

File: scope-frontend/src/utils/contentSafety.ts
const LEET_REPLACEMENTS: Record<string, string> = {
  '0': 'o',
  '1': 'i',
  '3': 'e',
  '4': 'a',
  '5': 's',
  '7': 't',
  '@': 'a',
  '$': 's',
};

const BLOCKED_TERMS = new Set([
  'chink',
  'coon',
  'fag',
  'faggot',
  'gook',
  'kike',
  'kill all',
  'kill yourself',
  'kys',
  'nigga',
  'nigger',
  'rape',
  'retard',
  'spic',
  'tranny',
  'wetback',
  'scope test blocked slur',
]);

const TOKEN_BLOCKED_TERMS = new Set([...BLOCKED_TERMS].filter((term) => !term.includes(' ')));

export interface ContentSafetyResult {
  clean: boolean;
  field?: string;
  message?: string;
}

function normalizeSafetyText(value: unknown): string {
  return String(value ?? '')
    .normalize('NFKD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .replace(/[013457@$]/g, (match) => LEET_REPLACEMENTS[match] ?? match);
}

function tokenize(value: unknown): string[] {
  return normalizeSafetyText(value).split(/[^a-z0-9]+/).filter(Boolean);
}

function containsBlockedTerm(value: unknown): boolean {
  const tokens = tokenize(value);
  if (!tokens.length) {
    return false;
  }

  if (tokens.some((token) => TOKEN_BLOCKED_TERMS.has(token))) {
    return true;
  }

  const phrase = tokens.join(' ');
  return [...BLOCKED_TERMS].some((term) => term.includes(' ') && phrase.includes(term));
}

export function validateContentSafety(fields: Array<{ field: string; value: unknown }>): ContentSafetyResult {
  for (const entry of fields) {
    const values = Array.isArray(entry.value) ? entry.value : [entry.value];
    if (values.some((value) => containsBlockedTerm(value))) {
      return {
        clean: false,
        field: entry.field,
        message: 'This contains a blocked slur or hate term.',
      };
    }
  }

  return { clean: true };
}
